src/Plugin/facets/widget/DateWidget.php

```diff
@@ -226,6 +322,134 @@ protected function doBuildDateWidgetElements(FacetInterface $facet, array $form
+    $default_month = (!empty($first_date_default) ? (int) $first_date_default->format('m') : NULL);
+    if (empty($second_date_default)) {
+      $second_date_default = $this->getDateFromActiveFilters($facet, 'second');
+    }
+    $last_month = (!empty($second_date_default) ? (int) $second_date_default->format('m') : NULL);
+    if ($default_month !== $last_month) {
+      $default_month = NULL;
+    }
+    $options_year = ['' => $this->t('Year')];
+    foreach ($year_months as $year => $months) {
+      $options_year[$year] = $year;
+    }
+    if (empty($default_year) || empty($options_year[$default_year])) {
+      $default_year = NULL;
+    }
+    if (empty($default_year)) {
+      $default_month = NULL;
+    }
+    $year_element_name = $facet->id() . '_year';
+    $month_element_name = $facet->id() . '_month';
+    $build[$year_month_wrapper][$year_element_name] = [
+      '#type' => 'select',
+      '#attributes' => ['class' => ['oe-list-pages-date-widget-year']],
+      '#title' => $this->t('Year'),
+      '#options' => $options_year,
+      '#default_value' => $default_year,
+    ];
+    $year_selector = $year_month_wrapper . '[' . $year_element_name . ']';
+    if ($parents) {
+      $first_parent = array_shift($parents);
+      $element_names = [$year_month_wrapper, $year_element_name];
+      $year_selector = $first_parent . '[' . implode('][', array_merge($parents, $element_names)) . ']';
+    }
+    $options_month = DateHelper::monthNames();
+    $options_month[''] = $this->t('Month');
+    $build[$year_month_wrapper][$month_element_name] = [
+      '#type' => 'select',
+      '#attributes' => ['class' => ['oe-list-pages-date-widget-month']],
+      '#title' => $this->t('Month'),
+      '#options' => $options_month,
+      '#default_value' => $default_month,
+      '#states' => [
+        'visible' => [
+          [
+            ':input[name="' . $year_selector . '"]' => [
+              'filled' => TRUE,
+            ],
+          ],
+        ],
+      ],
+    ];
+    $build['#attached']['library'][] = 'oe_list_pages/date_widget';
+  }
+
+  /**
+   * Get list of years and months from data.
+   */
+  protected function getYearMonths(FacetInterface $facet) {
+    $year_months = [];
+    // Get the entity type id, bundle and field name for the query.
+    [, $entity_type_id, $bundle] = explode(':', $facet->getFacetSourceId());
+    $bundle_key = $this->entityTypeManager->getDefinition($entity_type_id)->getKey('bundle');
+    $field_name = $facet->getFieldIdentifier();
+
+    // Build the Search API query.
+    $index = $facet->getFacetSource()->getIndex();
+    $query = $index->query();
+    $fields = $index->getFields();
+    if (in_array($bundle_key, array_keys($fields))) {
+      $query->addCondition($bundle_key, $bundle);
+    }
+    $query->addCondition('search_api_datasource', 'entity:' . $entity_type_id);
+    $query->range(0, 100000);
```

Comment:
This is not the correct approach. Facets already returns value for the results in the main query. Running a second query, loading all the items in the index is not the solution.
Instead, using the "date item processor" from facets module will allow to return results with the correct granularity. If then a special rendering is needed (dependent dropdown), it can be achieved by creating a dedicated widget.

Unfortunately there's a bug at the moment when using this date processor, which causes the page to crash.
This happens because the facets rebuild procedure doesn't take into account some data processing like the date item processor does.